File: src/syucream2/closure4.scala
import java.io.{ByteArrayOutputStream, ObjectOutputStream}
object Closure4 {
  def main(args: Array[String]) {
    val myClosure = new MyClosure
    DefaultSerializer.serialize(myClosure.serializable)  // => passed!
    DefaultSerializer.serialize(myClosure.notSerializable)  // => failed...
} }
class MyClosure {
  import MyClosure._
  private val value = 1
  def notSerializable = () => value
  def serializable = () => valueInObject
}
object MyClosure {
  val valueInObject = 1
}
object DefaultSerializer {
  def serialize(data: AnyRef): Array[Byte] = {
    val buffer = new ByteArrayOutputStream()
    val outStream = new ObjectOutputStream(buffer)
    outStream.writeObject(data)
    buffer.toByteArray
  }
}

File: src/syucream2/Closure4.java
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.ObjectOutputStream;
import java.io.Serializable;
import java.util.function.Supplier;

class Closure4 {
  public static void main(String[] args) {
    MyClosure myClosure = new MyClosure();

    try {
      DefaultSerializer.serialize(myClosure.serializable());
      DefaultSerializer.serialize(myClosure.notSerializable());
    } catch (Exception e) {
      e.printStackTrace();
    }
  }
}

interface SerializableSupplier<T> extends Supplier<T>, Serializable {};

class MyClosure {
  private final int value = 1;

  public SerializableSupplier<Integer> notSerializable() {
    return () -> {
      return this.value;
    };
  }

  public SerializableSupplier<Integer> serializable() {
    return () -> {
      return MyClosureObject.getInstance().valueInObject();
    };
  }
}

final class MyClosureObject {
  private static MyClosureObject instance;
  private int valueInObject = 1;

  static {
    instance = new MyClosureObject();
  }

  public static MyClosureObject getInstance() {
    return instance;
  }

  public int valueInObject() {
    return instance.valueInObject;
  }
}

class DefaultSerializer {
  public static byte[] serialize(Object data) throws IOException {
    ByteArrayOutputStream buffer = new ByteArrayOutputStream();
    ObjectOutputStream outStream = new ObjectOutputStream(buffer);
    outStream.writeObject(data);

    return buffer.toByteArray();
  }
}
